File: proposed-projects/round-trip-experiment/reverse/20260909T125752Z_09b2c811/20260910T072624Z_d6bd66eb/ported_implementation/gbnf/grammar_graph/pointers.py
"""An id keyed collection of graph pointers."""

from typing import Generator

from .graph_pointer import GraphPointer


class Pointers:
    def __init__(self, *pointers: GraphPointer):
        self.pointers = {}
        for pointer in pointers:
            self.add(pointer)

    def add(self, pointer: GraphPointer) -> None:
        self.pointers[pointer.id] = pointer

    def __iter__(self) -> Generator[GraphPointer, None, None]:
        yield from list(self.pointers.values())

    @property
    def size(self) -> int:
        return len(self.pointers)

    def __len__(self) -> int:
        return len(self.pointers)

File: packages/gbnf-experiment/data/20260909T001426Z_1c28aa33/ported_implementation/src/grammar-graph/pointers.ts
import type { GraphPointer } from './graph-pointer';

export class Pointers {
  #pointers = new Map<string, GraphPointer>();

  constructor(...pointers: GraphPointer[]) {
    for (const pointer of pointers) {
      this.add(pointer);
    }
  }

  add(pointer: GraphPointer): void {
    this.#pointers.set(pointer.id, pointer);
  }

  *[Symbol.iterator](): Generator<GraphPointer> {
    yield* this.#pointers.values();
  }

  get size(): number {
    return this.#pointers.size;
  }
}
